presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushRemoteExchangeThroughGroupId.java

```diff
@@ -0,0 +1,193 @@
+ */
+public final class PushRemoteExchangeThroughGroupId
+        implements Rule<ExchangeNode>
+{
+    private final Metadata metadata;
+    private static final Capture<GroupIdNode> GROUP_ID = newCapture();
+    private static final Pattern<ExchangeNode> PATTERN = exchange()
+            .matching(exchange -> exchange.getScope().isRemote())
+            .matching(exchange -> exchange.getType() == REPARTITION)
+            .with(source().matching(
+                    groupId()
+                            .capturedAs(GROUP_ID)
+                            .matching(groupId -> !groupId.getCommonGroupingColumns().isEmpty())));
+
+    public PushRemoteExchangeThroughGroupId(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<ExchangeNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return shouldPushRemoteExchangeThroughGroupId(session);
+    }
+
+    @Override
+    public Result apply(ExchangeNode node, Captures captures, Context context)
+    {
+        GroupIdNode groupIdNode = captures.get(GROUP_ID);
+
+        List<VariableReferenceExpression> inputs = getOnlyElement(node.getInputs());
+        inputs = removeVariable(inputs, groupIdNode.getGroupIdVariable());
+        inputs = replaceAlias(inputs, groupIdNode.getGroupingColumns());
+
+        PartitioningScheme partitioningScheme = node.getPartitioningScheme();
+        List<VariableReferenceExpression> outputLayout = partitioningScheme.getOutputLayout();
+        outputLayout = removeVariable(outputLayout, groupIdNode.getGroupIdVariable());
+        outputLayout = replaceAlias(outputLayout, groupIdNode.getGroupingColumns());
+
+        Set<VariableReferenceExpression> commonGroupingColumns = groupIdNode.getCommonGroupingColumns();
+        List<VariableReferenceExpression> partitionColumns = replaceAlias(commonGroupingColumns, groupIdNode.getGroupingColumns());
+
+        // Check that new |partitionColumns| must be subset of original partition columns.
+        Map<VariableReferenceExpression, VariableReferenceExpression> groupingColumns = groupIdNode.getGroupingColumns();
+        List<VariableReferenceExpression> originalPartitionColumns =
+                partitioningScheme.getPartitioning().getVariableReferences()
+                        .stream()
+                        .map(expr -> groupingColumns.getOrDefault(expr, expr))
+                        .collect(toImmutableList());
+        if (!originalPartitionColumns.containsAll(partitionColumns)) {
+            return Result.empty();
+        }
+
+        // Create new PartitioningHandle for hive.
+        PartitioningHandle partitioningHandle;
+        String catalog = getPartitioningProviderCatalog(context.getSession());
+        if (catalog.equals("hive")) {
+            partitioningHandle = createPartitioningHandle(context.getSession(), partitionColumns);
```

Comment:
this won't work, as you can name your "hive" catalog anything you want, and you can also have other custom catalog support (e.g. internally we use prism_batch).

Can we instead follow the same logic as in addExchanges to exclude the system catalog:
 `if (GlobalSystemConnector.NAME.equals(partitioningProviderCatalog))`